### PR TITLE
More flexible hook system

### LINKED ISSUE #1754: Hook to act after others plugins invocation
### What problem does this feature solve?
(Plugins dev API)

I don't know if this already exists or possible, but I need a method to listen other plugins 
invocation inside my plugin, to alter files contents.

Simple use case:

I want to replace default import of PWA plugin (`import './registerServiceWorker'`) to :

```js
if (process.client) require('./registerServiceWorker');
```

ATM, I use `api.postProcessFiles()` function, but this one only work if I do:

```bash
vue add @vue/pwa
vue add my-plugin
```

But if I do

```bash
vue add my-plugin
vue add @vue/pwa
```

`my-plugin` is not aware that PWA was added/invoked, so I cant transform the import statement.

### What does the proposed API look like?
Something like `api.postProcessFiles()`

```js
api.hookPluginInvoke('pluginName', (files) => {})
```

Thanks !

### LINKED ISSUE #1938: Plugin depending on another plugin
### What problem does this feature solve?
Some plugins might want to depend on other plugins. Currently, there is no way to do this.

### What does the proposed API look like?
If the `package.json` `dependencies` of `vue-cli-plugin-a` contains `vue-cli-plugin-b`, then:

1. `vue add a` will run `vue add b` and then `vue add a`
2. `vue invoke a` will run `vue invoke b` and then `vue invoke a` (Optionally configured by `a`)

We can do this multiple deep levels.

### Issues

1. If `a` depends on `b` and `c`, what's the order of `b` and `c` to be run?
2. What about circular dependencies?

### PR DESCRIPTION
**Pull request by pksunkara**
Fixes #1754, Fixes #1938

Implements the discussion regargding hooks from https://gist.github.com/sodatea/37bc8c36d7d0c285dfea52fbe8ddaf90

### CODE REVIEW

**Comment by pksunkara** [comment]
@sodatea Can I get some feedback on this?

**Comment by sodatea** [comment]
I'm currently working on a patch release.
Will look into this later for the next minor version.

**Comment by LinusBorg** [comment]
@sodatea Is this still on the rosdmap in your opinion?

**Comment by sodatea** [comment]
@LinusBorg Yeah I’ve talked to Evan before. He may have some additional ideas on the API design.

File: packages/@vue/cli/lib/Generator.js [hunk 1]
```diff
@@ -242,12 +275,22 @@ module.exports = class Generator {
...
+    ].some(id => {
+      if (!matchesPluginId(_id, id)) {
+        return false
+      }
+
+      if (!_version) {
+        return true
+      }
+
+      const version = this.pkg.dependencies[id] || this.pkg.devDependencies[id]
```
Comment (sodatea) [inline packages/@vue/cli/lib/Generator.js]: This typically returns a semver range like `^3.9.0` and is not a valid version to be used as the first argument of `semver.satisfies`.
Better use [getInstalledVersion](https://github.com/vuejs/vue-cli/blob/4af27247ccabe2ba5da57b6f216d17ad0767b8a8/packages/%40vue/cli/lib/util/ProjectPackageManager.js#L151) and do some workaround in the unit tests.

File: packages/@vue/cli/lib/GeneratorAPI.js [hunk 2]
```diff
@@ -133,10 +133,11 @@ class GeneratorAPI {
    * Check if the project has a given plugin.
    *
    * @param {string} id - Plugin id, can omit the (@vue/|vue-|@scope/vue)-cli-plugin- prefix
+   * @param {string} version - Plugin version. Defaults to ''
    * @return {boolean}
    */
-  hasPlugin (id) {
-    return this.generator.hasPlugin(id)
+  hasPlugin (id, version = '') {
```
Comment (sodatea) [inline packages/@vue/cli/lib/GeneratorAPI.js]: I think `undefined` is fine here…

File: packages/@vue/cli/lib/add.js [hunk 3]
```diff
@@ -27,12 +26,7 @@ async function add (pluginName, options = {}, context = process.cwd()) {
   log(`${chalk.green('✔')}  Successfully installed plugin: ${chalk.cyan(packageName)}`)
   log()
 
-  const generatorPath = resolveModule(`${packageName}/generator`, context)
-  if (generatorPath) {
-    invoke(pluginName, options, context)
-  } else {
-    log(`Plugin ${packageName} does not have a generator to invoke`)
```
Comment (sodatea) [inline packages/@vue/cli/lib/add.js]: Why remove this?

File: packages/@vue/cli/lib/Generator.js [hunk 4]
```diff
@@ -93,14 +95,45 @@ module.exports = class Generator {
...
     const rootOptions = cliService
       ? cliService.options
       : inferRootOptions(pkg)
+
+    // apply hooks from all plugins
+    this.otherPlugins.forEach(id => {
+      const api = new GeneratorAPI(id, this, {}, rootOptions)
+      const pluginGenerator = loadModule(`${id}/generator`, context)
+      if (pluginGenerator && pluginGenerator.hooks) {
+        pluginGenerator.hooks(api, {}, rootOptions, pluginIds)
```
Comment (sodatea) [inline packages/@vue/cli/lib/Generator.js]: With this design & implementation, `afterInvoke` callbacks from other plugins also gets pushed into the callback arrays, this is not the expected behavior.
I think the `hooks` feature should be more restrictive. Object syntax may be a better alternative.

File: packages/@vue/cli/lib/add.js (hunk 3 above)
Comment (pksunkara) [reply packages/@vue/cli/lib/add.js]: This is just an useless duplicate. `invoke` already does the same check at https://github.com/pksunkara/vue-cli/blob/hook/packages/%40vue/cli/lib/invoke.js#L64 pretty early.

Comment (sodatea) [reply packages/@vue/cli/lib/add.js]: Ah that's true, thanks.

File: packages/@vue/cli/lib/Generator.js (hunk 4 above)
Comment (pksunkara) [reply packages/@vue/cli/lib/Generator.js]: As you can see here, I am resetting the non-any hooks. https://github.com/vuejs/vue-cli/pull/2337/files#diff-f0e2bfadbc7f0f35b5cc089613387802R126.

Comment (sodatea) [reply packages/@vue/cli/lib/Generator.js]: Okay.

File: packages/@vue/cli/lib/invoke.js [hunk 5]
```diff
@@ -132,9 +135,18 @@ async function runGenerator (context, plugin, pkg = getPkg(context)) {
...
-    for (const cb of createCompleteCbs) {
+    for (const cb of afterInvokeCbs) {
+      await cb()
+    }
+    stopSpinner()
+    log()
+  }
+
+  if (afterAnyInvokeCbs.length) {
+    logWithSpinner('⚓', `Running completion hooks from other plugins...`)
```
Comment (sodatea) [inline packages/@vue/cli/lib/invoke.js]: Does this include the `afterAnyInvoke` hook from the currently running generator?

Comment (pksunkara) [reply packages/@vue/cli/lib/invoke.js]: Yes, it does.

Comment (pksunkara) [reply packages/@vue/cli/lib/invoke.js]: From all the current plugins.

**Comment by pksunkara** [comment]
Updated the PR

File: packages/@vue/cli/lib/invoke.js (hunk 5 above)
Comment (sodatea) [reply packages/@vue/cli/lib/invoke.js]: But the order of `afterAnyInvoke` callbacks is first all other plugins' hook, then the current plugin's `afterAnyInvoke` hook.

The exact result depends on which plugin did you invoke first.
This looks confusing to me.
It is the same kind of non-determinism like `npm install`. We should avoid it.

Comment (sodatea) [reply packages/@vue/cli/lib/invoke.js]: Also the log is misleading in that case.

Comment (pksunkara) [reply packages/@vue/cli/lib/invoke.js]: Fixed this. Current plugins will run first and then others. The order is now documented.

Comment (sodatea) [reply packages/@vue/cli/lib/invoke.js]: It no longer triggers other plugins' `afterAnyInvoke` hook as I tested.

I'm also a little bit skeptical about this approach - this is still non-deterministic IMO.
For example, if a user has two plugins with `afterAnyInvoke` hook, say, `eslint` and `typescript`, the order will be:

- `eslint` -> `typescript` on creation
- `eslint` -> `typescript` if user add `eslint` first
- `typescript` -> `eslint` if user add `typescript` first

My idea is to keep the order of `afterAnyInvoke` in sync with that of plugin loading (for now, it is the key enumeration order in `package.json`) and then figure out a way to specify plugin order in a separate PR. What do you think?

Comment (pksunkara) [reply packages/@vue/cli/lib/invoke.js]: Good idea. I didn't think of it like that. Made it completely deterministic like how you said. Should be working now.

**Pull request merged**